tests/core/data/test_tensordense.py

```diff
@@ -0,0 +1,234 @@
+        pytest.param(_valid_numpy(), {'shape': ()}, ValueError,
+                     id="numpy-shape 0 tuple"),
+        pytest.param(_valid_numpy(), {'shape': (1,)}, ValueError,
+                     id="numpy-shape 1 tuple"),
+        pytest.param(_valid_numpy(), {'shape': (None, None)}, ValueError,
+                     id="numpy-shape None tuple"),
+        pytest.param(_valid_numpy(), {'shape': [2, 2]}, ValueError,
+                     id="numpy-shape list"),
+        pytest.param(_valid_numpy(), {'shape': (1, 2, 3)}, ValueError,
+                     id="numpy-shape 3 tuple"),
+        pytest.param(_valid_numpy(), {'shape': (-1, 1)}, ValueError,
+                     id="numpy-negative shape"),
+        pytest.param(_valid_numpy(), {'shape': (-4, -4)}, ValueError,
+                     id="numpy-both negative shape"),
+        pytest.param(_valid_numpy(), {'shape': (1213, 1217)}, ValueError,
+                     id="numpy-different shape"),
+        pytest.param(_valid_list(), {'shape': ()}, ValueError,
+                     id="numpy-shape 0 tuple"),
+        pytest.param(_valid_list(), {'shape': (1,)}, ValueError,
+                     id="numpy-shape 1 tuple"),
+        pytest.param(_valid_list(), {'shape': (None, None)}, ValueError,
+                     id="numpy-shape None tuple"),
+        pytest.param(_valid_list(), {'shape': [2, 2]}, ValueError,
+                     id="numpy-shape list"),
+        pytest.param(_valid_list(), {'shape': (1, 2, 3)}, ValueError,
+                     id="numpy-shape 3 tuple"),
+        pytest.param(_valid_list(), {'shape': (-1, 1)}, ValueError,
+                     id="numpy-negative shape"),
+        pytest.param(_valid_list(), {'shape': (-4, -4)}, ValueError,
+                     id="numpy-both negative shape"),
+        pytest.param(_valid_list(), {'shape': (1213, 1217)}, ValueError,
+                     id="numpy-different shape"),
+        pytest.param(_valid_tensor(), {'shape': ()}, ValueError,
+                     id="numpy-shape 0 tuple"),
+        pytest.param(_valid_tensor(), {'shape': (1,)}, ValueError,
+                     id="numpy-shape 1 tuple"),
+        pytest.param(_valid_tensor(), {'shape': (None, None)}, ValueError,
+                     id="numpy-shape None tuple"),
+        pytest.param(_valid_tensor(), {'shape': [2, 2]}, ValueError,
+                     id="numpy-shape list"),
+        pytest.param(_valid_tensor(), {'shape': (1, 2, 3)}, ValueError,
+                     id="numpy-shape 3 tuple"),
+        pytest.param(_valid_tensor(), {'shape': (-1, 1)}, ValueError,
+                     id="numpy-negative shape"),
+        pytest.param(_valid_tensor(), {'shape': (-4, -4)}, ValueError,
+                     id="numpy-both negative shape"),
+        pytest.param(_valid_tensor(), {'shape': (1213, 1217)}, ValueError,
+                     id="numpy-different shape"),
+    ])
+    def test_init_from_wrong_input(self, arg, kwargs, error):
```

Comment:
This parametrisation is too long...
All error are value error, thus it it not needed.
You are also testing all tensor type ('arg') with each shape.
You could parametrize them independently. Also so `id`s are repeated.

Reply:
I think now should be better.